exonum-java-binding-core/src/main/java/com/exonum/binding/storage/indices/ListSpliterator.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.storage.indices;
+
+import static com.exonum.binding.storage.indices.StoragePreconditions.checkPositionIndex;
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.exonum.binding.storage.database.ModificationCounter;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ConcurrentModificationException;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+/**
+ * A spliterator for list-like Exonum collections.
+ *
+ * <p>This spliterator is late-binding and fail-fast. It will fail if <em>any</em> collection
+ * associated with a given {@link com.exonum.binding.storage.database.Fork} is modified; not only
+ * the source list.
+ *
+ * <p>This spliterator does not support specializations (e.g., {@link Spliterator.OfInt}).
+ * If they are ever needed, see the Spliterator in an archived "exonum-serialization" project.
+ *
+ * @param <ElementT> the type of elements in the corresponding list this spliterator provides
+ */
+/*
+ * todo: Can’t we make a spliterator that is resilient to modifications of unrelated collections?
+ *  I think we can as we use indices into list.
+ *
+ * todo: Shall we use the iterator (from a certain index) that Rust can provide? Is it more
+ *   efficient?
+ */
+class ListSpliterator<ElementT> implements Spliterator<ElementT> {
+
+  /** Characteristics of any list spliterator. */
+  private static final int ANY_CHARACTERISTICS = NONNULL | ORDERED | SIZED | SUBSIZED;
+
+  @VisibleForTesting
+  static final int MIN_SPLITTABLE_SIZE = 2;
+
+  private final ListIndex<ElementT> list;
+
+  /** An index of the next element to read. */
+  private long nextIndex;
+
+  /** After-the-last index. */
+  private long fence;
+
+  private final ModificationCounter counter;
+
+  private final int characteristics;
+
+  /**
+   * The value of the modification counter at the bind time — the start of iteration.
+   */
+  private Integer initialCounterValue;
+
+  ListSpliterator(ListIndex<ElementT> list, ModificationCounter counter, boolean immutable) {
```

Comment:
Same here - can't we have just one argument `immutable` and infer the type of `ModificationCounter` from it?

Reply:
What do you mean by "infer"? The modification counter comes from a view, and is used to detect changes made to it. We cannot instantiate an independent counter here.

Reply:
Sorry, I meant that we could look at the type of the counter (whether it is mutable or not) and get rid of an `immutable` constructor parameter (same in the case Fork/Snapshot). But if removing an argument and making this class instantiation more straightforward for the user not worth putting conditional logic into the constructor for then it's good to me too

Reply:
But there isn't an `isMutable` method in `ModificationCounter`, is it? Re ease of instantiation, it is an internal class.

Reply:
Yes, I meant adding `isMutable` in `ModificationCounter` :) But I agree that it's too much work for an internal class so it's good 👍